source/lac/trilinos_sparsity_pattern.cc

```diff
@@ -114,43 +114,35 @@ namespace TrilinosWrappers
     void
     Accessor::visit_present_row ()
     {
-      // if we are asked to visit the
-      // past-the-end line, then simply
-      // release all our caches and go on
-      // with life
+      // if we are asked to visit the past-the-end line, then simply
+      // release all our caches and go on with life
       if (this->a_row == sparsity_pattern->n_rows())
         {
           colnum_cache.reset ();
 
           return;
         }
-//TODO: Is this thread safe?
 
       // otherwise first flush Trilinos caches
       sparsity_pattern->compress ();
 
-      // get a representation of the present
-      // row
+      // get a representation of the present row
       int ncols;
-      // TODO: casting a size_type to an int, could be a problem
-      int colnums = sparsity_pattern->n_cols();
-
-      int ierr;
-      ierr = sparsity_pattern->graph->ExtractGlobalRowCopy((TrilinosWrappers::types::int_type)this->a_row,
-                                                           colnums,
-                                                           ncols,
-                                                           (TrilinosWrappers::types::int_type *)&(*colnum_cache)[0]);
-      AssertThrow (ierr == 0, ExcTrilinosError(ierr));
 
-      // copy it into our caches if the
-      // line isn't empty. if it is, then
-      // we've done something wrong, since
-      // we shouldn't have initialized an
-      // iterator for an empty line (what
-      // would it point to?)
-      Assert (ncols != 0, ExcInternalError());
-      colnum_cache.reset (new std::vector<size_type> (colnums,
-                                                      colnums+ncols));
+      colnum_cache.reset (new std::vector<size_type> (sparsity_pattern->row_length(this->a_row)));
```

Comment:
There is a comment in the class that says that we want to make copying of accessors cheap. Consequently, this cache is administered as a `shared_ptr`. If you just resize things, then other iterators/accessors that were copied from the current one before moving to a new row, will become outdated.

The `_02` test I added checks for exactly this.

I think one could debate whether copying iterators is a case that needs to be supported in a particularly efficient way. That says, the class currently supports it, and I'm not particularly worried about applications that walk over the elements of a sparsity pattern -- that's mostly for debugging, so I didn't bother with trying to improve efficiency over what it currently is. My interest was mostly in just making it work.